File: aarrepo/src/main/java/com/zaze/aarrepo/utils/DeviceUtil.java
package com.zaze.aarrepo.utils;

import android.content.Context;
import android.os.Build;
import android.os.Environment;
import android.os.StatFs;
import android.provider.Settings;
import android.telephony.TelephonyManager;

import com.zaze.aarrepo.commons.log.LogKit;

import java.io.DataOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;
import java.util.UUID;

/**
 * Description : 设备工具类
 * date : 2015-11-27 - 17:11
 *
 * @author : zaze
 * @version : 1.0
 */
public class DeviceUtil {

    /**
     * @return 检查设备是否Root了
     */
    public static boolean checksRoot() {
        Process process = null;
        DataOutputStream outputStream = null;
        try {
            process = Runtime.getRuntime().exec("su");
            outputStream = new DataOutputStream(process.getOutputStream());
            outputStream.writeBytes("exit\n");
            outputStream.flush();
            int exitValue = process.waitFor();
            if (exitValue == 0) {
                LogKit.i("设备已Root");
                return true;
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (outputStream != null) {
                try {
                    outputStream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (process != null) {
                process.destroy();
            }
        }
        LogKit.i("设备未Root");
        return false;
    }

    /**
     * @param context
     * @return 机器码
     */
    public static String getMachineCode(Context context) {
        UUID uuid;
        String androidId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
        try {
            if (!"9774d56d682e549c".equals(androidId)) {
                uuid = UUID.nameUUIDFromBytes(androidId.getBytes("utf8"));
            } else {
                final String deviceId = ((TelephonyManager) context
                        .getSystemService(Context.TELEPHONY_SERVICE))
                        .getDeviceId();
                uuid = deviceId != null ? UUID.nameUUIDFromBytes(deviceId
                        .getBytes("utf8")) : UUID.randomUUID();
            }
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
        return uuid.toString();
    }

    /**
     * @return 机器型号
     */
    public static String getModel() {
        return Build.MODEL;
    }

    /**
     * @return 系统版本
     */
    public static String getSysVersion() {
        return Build.VERSION.RELEASE;
    }

    /**
     * @return 本地mac 地址
     */
    public static String getMacAddress() {
 /*获取mac地址有一点需要注意的就是android 6.0版本后，以下注释方法不再适用，不管任何手机都会返回"02:00:00:00:00:00"这个默认的mac地址，这是googel官方为了加强权限管理而禁用了getSYstemService(Context.WIFI_SERVICE)方法来获得mac地址。*/
        //        String macAddress= "";
//        WifiManager wifiManager = (WifiManager) MyApp.getContext().getSystemService(Context.WIFI_SERVICE);
//        WifiInfo wifiInfo = wifiManager.getConnectionInfo();
//        macAddress = wifiInfo.getMacAddress();
//        return macAddress;

        String macAddress = null;
        StringBuffer buf = new StringBuffer();
        NetworkInterface networkInterface = null;
        try {
            networkInterface = NetworkInterface.getByName("eth1");
            if (networkInterface == null) {
                networkInterface = NetworkInterface.getByName("wlan0");
            }
            if (networkInterface == null) {
                return "02:00:00:00:00:02";
            }
            byte[] addr = networkInterface.getHardwareAddress();
            for (byte b : addr) {
                buf.append(String.format("%02X:", b));
            }
            if (buf.length() > 0) {
                buf.deleteCharAt(buf.length() - 1);
            }
            macAddress = buf.toString();
        } catch (SocketException e) {
            e.printStackTrace();
            return "02:00:00:00:00:02";
        }
        return macAddress;
    }

    /**
     * @return sdcard空间大小
     */
    public static long getSDTotalSpace() {
        File sd = Environment.getExternalStorageDirectory();
        StatFs stat = new StatFs(sd.getPath());
        long blockSize = stat.getBlockSize();
        long totalBlocks = stat.getBlockCount();
        return blockSize * totalBlocks;
    }

    /**
     * @return sdcard 剩余空间大小
     */
    public static long getSDFreeSpace() {
        File sd = Environment.getExternalStorageDirectory();
        StatFs stat = new StatFs(sd.getPath());
        long blockSize = stat.getBlockSize();
        long availableBlocks = stat.getAvailableBlocks();
        return blockSize * availableBlocks;
    }

    /**
     * ROM
     *
     * @return 获取机身总大小
     */
    public static long getRomTotalSpace() {
        File rom = Environment.getDataDirectory();
        StatFs stat = new StatFs(rom.getPath());
        long blockSize = stat.getBlockSize();
        long totalBlocks = stat.getBlockCount();
        return blockSize * totalBlocks;
    }

    /**
     * ROM
     *
     * @return 获取机身剩余
     */
    public static long getRomFreeSpace() {
        File rom = Environment.getDataDirectory();
        StatFs stat = new StatFs(rom.getPath());
        long blockSize = stat.getBlockSize();
        long totalBlocks = stat.getAvailableBlocks();
        return blockSize * totalBlocks;
    }


    /**
     * RAM
     * java 虚拟机 最大内存
     *
     * @return
     */
    public static long getVMMaxMemory() {
        return Runtime.getRuntime().maxMemory();
    }

    /**
     * RAM
     * java 虚拟机 当前 从机器内存中取过来的 内存的 中的空闲内存
     *
     * @return
     */
    public static long getVMFreeMemory() {
        return Runtime.getRuntime().freeMemory();
    }

    /**
     * RAM
     * java 虚拟机 当前 从机器内存中取过来 的 总内存(包括使用了的和 freeMemory)
     *
     * @return
     */
    public static long getVMTotalMemory() {
        return Runtime.getRuntime().totalMemory();
    }

    public String getLocalIpAddress() {
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); en.hasMoreElements(); ) {
                NetworkInterface intf = en.nextElement();
                for (Enumeration<InetAddress> enumIpAddr = intf.getInetAddresses(); enumIpAddr.hasMoreElements(); ) {
                    InetAddress inetAddress = enumIpAddr.nextElement();
                    if (!inetAddress.isLoopbackAddress()) {
                        return inetAddress.getHostAddress().toString();
                    }
                }
            }
        } catch (SocketException ex) {
            ex.printStackTrace();
        }
        return null;
    }

    /**
     * @param context
     * 测试打印
     */
//    public static void printPhoneMsg(Context context) {
//        TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
//        StringBuilder sb = new StringBuilder();
//        sb.append("\nDeviceId(IMEI) = " + tm.getDeviceId());
//        sb.append("\nDeviceSoftwareVersion = " + tm.getDeviceSoftwareVersion());
//        sb.append("\nLine1Number = " + tm.getLine1Number());
//        sb.append("\nNetworkCountryIso = " + tm.getNetworkCountryIso());
//        sb.append("\nNetworkOperator = " + tm.getNetworkOperator());
//        sb.append("\nNetworkOperatorName = " + tm.getNetworkOperatorName());
//        sb.append("\nNetworkType = " + tm.getNetworkType());
//        sb.append("\nPhoneType = " + tm.getPhoneType());
//        sb.append("\nSimCountryIso = " + tm.getSimCountryIso());
//        sb.append("\nSimOperator = " + tm.getSimOperator());
//        sb.append("\nSimOperatorName = " + tm.getSimOperatorName());
//        sb.append("\nSimSerialNumber = " + tm.getSimSerialNumber());
//        sb.append("\nSimState = " + tm.getSimState());
//        sb.append("\nSubscriberId(IMSI) = " + tm.getSubscriberId());
//        sb.append("\nVoiceMailNumber = " + tm.getVoiceMailNumber());
//        sb.append("\nProduct Model: " + android.os.Build.MODEL + ","
//                + android.os.Build.VERSION.SDK + ","
//                + android.os.Build.VERSION.RELEASE);
//        LogDevelopmentKit.e("zaze 0720 : " + sb.toString());
//    }
}

File: zazeutil/src/main/java/com/zaze/utils/DeviceUtil.kt
package com.zaze.utils

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.content.Context
import android.content.Context.ACTIVITY_SERVICE
import android.os.Build
import android.os.Environment
import android.provider.Settings
import android.telephony.TelephonyManager
import android.text.TextUtils
import java.util.*


/**
 * Description :

 * @author : ZAZE
 * *
 * @version : 2017-08-01 - 13:43
 */
object DeviceUtil {
    // --------------------------------------------------
    // --------------------------------------------------
    /**
     * Description : 设备第一次启动时产生和存储的64bit的一个数，当设备被wipe后该数重置
     * @author zaze
     * @version 2017/8/1 - 下午1:52 1.0
     */
    @JvmStatic
    fun getAndroidId(context: Context): String {
        return Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
    }

    @SuppressLint("MissingPermission")
    @JvmStatic
    fun getDeviceId(context: Context): String? {
        return (context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager).deviceId
    }

    @SuppressLint("MissingPermission")
    @JvmStatic
    fun getSimSerialNumber(context: Context): String? {
        return (context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager).simSerialNumber
    }

    /**
     * SimSerialNumber -> DeviceId(IMEI) -> AndroidId -> randomId
     * [context] context
     */
    @JvmStatic
    fun getUUID(context: Context): String {
        val key = "getUUID"
        var id = Build.SERIAL
        if (TextUtils.isEmpty(id)) {
            id = getSimSerialNumber(context)
            if (TextUtils.isEmpty(id)) {
                id = getDeviceId(context)
                if (TextUtils.isEmpty(id)) {
                    id = getAndroidId(context)
                    if ("9774d56d682e549c" == id) {
                        val sharedPrefUtil = SharedPrefUtil.newInstance(context)
                        id = sharedPrefUtil.get(key, "")
                        if (TextUtils.isEmpty(id)) {
                            id = UUID.randomUUID().toString()
                            sharedPrefUtil.commit(key, id)
                        }
                    }
                }
            }
        }
        return id
    }

// --------------------------------------------------
// --------------------------------------------------
    /**
     * @return sdcard总空间大小
     */
    @JvmStatic
    fun getSdTotalSpace(): Long {
        return FileUtil.getTotalSpace(Environment.getExternalStorageDirectory())
    }

    /**
     * @return sdcard 剩余空间大小
     */
    @JvmStatic
    fun getSdFreeSpace(): Long {
        return FileUtil.getFreeSpace(Environment.getExternalStorageDirectory())
    }
// --------------------------------------------------
    /**
     * Data (内置sd卡 时 同 {@link #getSDTotalSpace()})
     * @return 获取机身总大小
     */
    @JvmStatic
    fun getDataTotalSpace(): Long {
        return FileUtil.getTotalSpace(Environment.getDataDirectory())
    }

    /**
     * Data
     * @return 获取机身剩余
     */
    @JvmStatic
    fun getDataFreeSpace(): Long {
        return FileUtil.getFreeSpace(Environment.getDataDirectory())
    }

// --------------------------------------------------
    /**
     * Runtime Max Memory
     * 单个应用 最大运存
     * @return
     */
    @JvmStatic
    fun getRuntimeMaxMemory(): Long {
        return Runtime.getRuntime().maxMemory()
    }

    /**
     * Runtime Free Memory
     * 当前 从机器内存中取过来的 内存的 中的空闲内存
     * @return
     */
    @JvmStatic
    fun getRuntimeFreeMemory(): Long {
        return Runtime.getRuntime().freeMemory()
    }

    /**
     * Runtime Total Memory
     * 当前 从机器内存中取过来 的 总内存(包括使用了的和 freeMemory)
     * @return
     */
    @JvmStatic
    fun getRuntimeTotalMemory(): Long {
        return Runtime.getRuntime().totalMemory()
    }

    // --------------------------------------------------
    // --------------------------------------------------

    @JvmStatic
    fun getDeviceMemory(context: Context): ActivityManager.MemoryInfo {
        val am = context.getSystemService(ACTIVITY_SERVICE) as ActivityManager
        val outInfo = ActivityManager.MemoryInfo()
        am.getMemoryInfo(outInfo)
        return outInfo
    }

}
